### ISSUE TITLE
Feedback on Linear Model Chapter

### ISSUE THREAD

**Issue by lorentzenchr** [INRIA/scikit-learn-mooc#276]
This MOOC is really great. Below are a few suggestions after a first partial read.

Chapter [Linear regression using scikit-learn](https://inria.github.io/scikit-learn-mooc/python_scripts/linear_regression_in_sklearn.html):

- Mention that the strategy for finding the best parameters is called (ordinary) **least squares**.

[Linear regression with non-linear link between data and target](https://inria.github.io/scikit-learn-mooc/python_scripts/linear_regression_non_linear_link.html)

- The word *link* is unfortunate as it might get confused with the *link* function of GLMs.
- As soon as v1.0 of scikit-learn arrives, mention splines with `SplineTransformer`.
- Speeking about *interactions* is a bit misleading in a setting with only one single feature. But I would really love to see an example that specifies 2 interaction terms: Given x1 to x3, create x1, x2, x3, x1*x2, x1*x3 (not x2*x3 and not x**2 terms either).

[Regularization of linear regression model](https://inria.github.io/scikit-learn-mooc/python_scripts/linear_models_regularization.html)

- Remove convergence warnings of Ridge (other solver?)
- Maybe a word about categorical variables and the `drop` parameter in OHE?
- "Let’s consider the case where features have an identical data dispersion". The term "dispersion" seems unusual to me. How about "features live on the same scale/dimension like both are times measured in microseconds."
   Later on: "our model will boost the weights of features with small dispersion" => It is about larger/smaller (floating point/machine) numbers, not about dispersion.
- What about scaling of OHE categorical features?

[Main take-away](https://inria.github.io/scikit-learn-mooc/linear_models/linear_models_module_take_away.html)

- Again the word "link".
- Mention that (manual) feature engineering and feature selection is (by far) the greatest impact/tuning factor for linear models (and therefore against over- and underfitting).

**Comment by lesteve** [comment]
> This MOOC is really great. Below are a few suggestions after a first partial read.

Thanks for the kind words! It's great to hear, especially since we have a beta-test phase going right now and we are hearing about many many small things that could be better.

Thanks for taking the time to put together some feed-back as well, this is really valuable! Working on the MOOC I realized that reviewing pedagogical content is even harder than reviewing code and reviewing code is already quite hard.

Looking at the points you raised, I tried to categorise into different buckets. Let me know if you have any comments on this rough categorisation!

### Things I think we should aim to fix reasonably quickly (end-April?)
Let's say before end April for the public launch (scheduled mid-May):

- "link" wording: good point I thought even in English it was weird so maybe we could use relationship?
- interactions with a single feature => better wording? suggestions more than welcome!
- dispersion wording
- What about scaling of OHE categorical features? I guess you mean that scaling OHE categorical features is not needed in practice since they are all on the same scale? Is there something else I am missing? We could add a note. This is something we may want to mention as well in the "Predictive Modeling Pipeline"

### Nice to have probably for the public launch

- convergence warning of Ridge. I had some comments from people that thought they had done something wrong, like clicked on a wrong button, when they saw a red warning in Jupyter. Since we aim to be accessible without a strong technical background (noble aim but in practice not so easy) I think we should take this kind of things into acount. At the same time warning like this will sometimes happen so maybe adding a note is good enough.
- OHE and drop parameter? I'd say maybe adding a small note. I think we actually have some exercise in the Hyperparameter tuning that uses a drop parameter so we kind of mention it at one point very briefly (IIRC this is a parameter we do random search, it has no impact and it is probably a bit advanced at this point of the course so I would not say this is a great idea to have this exercise but this is another question ...) 
 
### The things I am not sure about

Mostly because I know very little about this ...

- example that specifies 2 interaction terms: Given x1 to x3, create x1, x2, x3, x1x2, x1x3 (not x2*x3 and not x**2 terms either).
- Mention that (manual) feature engineering and feature selection is (by far) the greatest impact/tuning factor for linear models (and therefore against over- and underfitting).
 
### For later
- SplineTransformer (we are planning to do some updated version of the MOOC some time in the future, maybe end 2021, but let's not get ahead too much of ourselves 😉)

**Comment by lorentzenchr** [comment]
Categoricals/OHE:
- Scaling OHE: Categorical data arise very frequently in practice and with the usual recommendation "normalize your data", you seldom find an answer what to do with your (OHE) categoricals. It would be nice to at least mention this issue, e.g. "do not rescale OHE".
- OHE and drop: I haven't read this part, but you should always choose to drop or not to drop yourself and not in hyperparameter search. General rule: If unpenalized, then drop one category/level. If penalized, do not drop.

@lesteve Things you are not so sure about: In my experience, manual feature engineering (and feature selection) is indeed one of most important steps with linear models. Firstly, it is short-sighted to always compare to GBTs and the like without considering that you could include splines and interaction terms in your linear model. These terms do usually not arise themselves, you have to put them there, manually. Secondly, with what else could you play around (and compare on a validation set / do hyperparameter search light)?! Sure, there is the penalty parameter, but scikit-learn does not provide the possibility for a different penalization per feature or feature group, nor does it give you a lot of options for penalties.

**Comment by GaelVaroquaux** [comment]
Hi Christian,

All this is correct, but we need to keep in mind that beginners are
already struggling with so many things. Our experience when teaching has
shown the importance of simplifying and adding very few new concepts in
each lesson.

**Comment by ogrisel** [comment]
> It would be nice to at least mention this issue, e.g. "do not rescale OHE".

This we could probably do easily.

**Comment by ogrisel** [comment]
> OHE and drop: I haven't read this part, but you should always choose to drop or not to drop yourself and not in hyperparameter search. General rule: If unpenalized, then drop one category/level. If penalized, do not drop.

I think we might want to simplify the first module and not mention dropping redundant OHE features at all because we always penalize.

**Comment by GaelVaroquaux** [comment]
> I think we might want to simplify the first module and not mention dropping redundant OHE features at all because we always penalize.

+1
